### PR TITLE
Android license detector in doctor, take two

### LINKED ISSUE #13584: flutter license check comes too late in the install process, doctor doesn't warn about it
I'm installing flutter on a brand-new mac. I ran into a surprising situation, which I think we can make much better for new users.

tldr: flutter doctor says everything is fine, however things aren't fine: I didn't accept my licenses.

Proposal: have doctor check if I accepted my licenses, prompt me to do it _before_ I run my first app.

(apparently we know if the user accepted the licenses, because we tell the user to run "flutter doctor --android-licenses)

Here's what I see before I run my first app:

```
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.2)
    • Android SDK at /Users/sethladd/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
```

However, here's what I'm told when I try to run my first app:

```
sethladd-macbookpro:my_first_app sethladd$ flutter run
Launching lib/main.dart on Android SDK built for x86 in debug mode...
Initializing gradle...                                9.6s
Resolving dependencies...                                -

* Error running Gradle:
Unable to download needed Android SDK components, as the following licenses have not been accepted:
Android SDK Build-Tools 25.0.3, Android SDK Platform 25

To resolve this, please run the following command in a Terminal:
flutter doctor --android-licenses
\sethladd-macbookpro:my_first_app sethladd$ flutter doctor --android-licenses
Warning: File /Users/sethladd/.android/repositories.cfg could not be loaded.    
5 of 6 SDK package licenses not accepted. 100% Computing updates...             
Review licenses that have not been accepted (y/N)? y

```

Sad user, I thought I was good to go, because I had the checkmark, but roadbump!

I think our users would much rather see this warning the first time they run `flutter doctor`, at least when/while they run `flutter doctor` as they are setting up their Android environment, because that's when they are getting their dependencies setup.

The first time I run, assuming doctor gives me a checkmark, my app should _just run_.

Thanks for taking a look!

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #13584, for real this time.  This fixes all known problems with #14535, detailed [here](https://github.com/flutter/flutter/issues/13584#issuecomment-366051100).

New unit tests are included to exercise the features of the new AnsiSpinner as well as the old _AnsiStatus (now AnsiStatus).

Additionally, this fixes another problem where flutter doctor --android-licenses wouldn't work on MacOS (_sdkManagerEnv wasn't being set properly) and adjusts the offset for AnsiStatus so that progress indicators are shifted right enough for longer lines on update packages.

Before:
before

After:
after

### CODE REVIEW

**Comment by jcollins-g** [comment]
Parameter changes blew up mockito tests in many places and it doesn't seem like a quick fix.  Will reopen once that's dealt with.

**Comment by jcollins-g** [comment]
I'm not going to land this until I get access to a Windows machine to test it on interactively, but otherwise this seems to work on Linux and Mac.  Ready for review once more.

File: packages/flutter_tools/lib/src/android/android_studio.dart [hunk 1]
```diff
@@ -178,14 +178,16 @@ class AndroidStudio implements Comparable<AndroidStudio> {
...
-              (AndroidStudio other) => other.directory == studio.directory);
-          studios.add(studio);
+    if (fs.directory(homeDirPath).existsSync()) {
+      for (FileSystemEntity entity in fs.directory(homeDirPath).listSync()) {
+        if (entity is Directory && entity.basename.startsWith('.AndroidStudio')) {
+          final AndroidStudio studio = new AndroidStudio.fromHomeDot(entity);
+          if (studio != null &&
+              !_hasStudioAt(studio.directory, newerThan: studio.version)) {
+            studios.removeWhere(
+                    (AndroidStudio other) => other.directory == studio.directory);
```
Comment (xster) [inline packages/flutter_tools/lib/src/android/android_studio.dart]: something's off with the formatting I think

File: packages/flutter_tools/lib/src/base/logger.dart [hunk 2]
```diff
@@ -11,6 +11,8 @@ import 'io.dart';
 import 'terminal.dart';
 import 'utils.dart';
 
+const int kDefaultStatusPadding = 59;
```
Comment (xster) [inline packages/flutter_tools/lib/src/base/logger.dart]: could you add 7 to lib/src/ios/mac.dart's lines 369 and 396? (we can probably make them relative vs this constant now)

File: packages/flutter_tools/lib/src/base/logger.dart [hunk 3]
```diff
@@ -280,57 +271,138 @@ enum _LogType {
...
   @override
+  /// Calls [onFinish].
   void cancel() {
-    onFinish();
-
-    if (!live)
-      return;
-    live = false;
-
-    print('\b ');
```
Comment (xster) [inline packages/flutter_tools/lib/src/base/logger.dart]: I think we should probably keep the original behaviour of having a new line after cancel.

**Review by xster — APPROVED** [review]
LG for the logger parts

File: packages/flutter_tools/lib/src/android/android_studio.dart (hunk 1 above)
Comment (jcollins-g) [reply packages/flutter_tools/lib/src/android/android_studio.dart]: done

File: packages/flutter_tools/lib/src/base/logger.dart (hunk 2 above)
Comment (jcollins-g) [reply packages/flutter_tools/lib/src/base/logger.dart]: done

File: packages/flutter_tools/lib/src/base/logger.dart (hunk 3 above)
Comment (jcollins-g) [reply packages/flutter_tools/lib/src/base/logger.dart]: done

**Comment by xster** [comment]
LGTM

Looks like there's an extra `-` character while building https://travis-ci.org/flutter/flutter/jobs/347012334#L723, but it might just be a Travis terminal weirdness.

**Comment by jcollins-g** [comment]
@xster : good catch, that's a bug.   Hopefully the last one; my windows tests suggest this will work fine.

**Review by tvolkert — APPROVED** [review]
LGTM

**Comment by xster** [comment]
Nice, LG, thanks!

**Pull request merged**